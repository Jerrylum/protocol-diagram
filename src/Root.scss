@use "style/Essential.scss";
@use "style/Modals.scss";

#root {
  overflow: hidden;

  --bg-body-color: #ffffff;
  --bg-default-color: #ffffff;
  --bg-paper-color: #f3f3f3;
  --text-primary-color: rgba(0, 0, 0, 0.87);
  --text-disabled-color: rgba(0, 0, 0, 0.6);
  --link-color: rgba(0, 0, 0, 0.87);
  --hover-color: rgba(0, 0, 0, 0.08);
  --focused-color: rgba(0, 0, 0, 0.12);
  background-color: var(--bg-body-color);

  > #root-container {
    display: flex;
    width: 100vw;
    width: 100svw;
    height: 100vh;
    height: 100svh;
    box-sizing: border-box;
  }
}
