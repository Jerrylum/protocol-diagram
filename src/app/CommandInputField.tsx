import { Box, Button, TextField } from "@mui/material";
import { observer } from "mobx-react-lite";

export const CommandInputField = observer(() => {
  return (
    <Box sx={{ position: "relative" }}>
      <TextField fullWidth size="small" spellCheck={false} />
      <Box
        sx={{
          position: "absolute",
          top: "0",
          bottom: "0",
          left: "calc(100% + 4px)",
          width: "300px",
          display: "flex",
          alignItems: "center",
          gap: "4px"
        }}>
        <Button>Export As Text</Button>
        {/* <Button>Share URL</Button> */}
      </Box>
    </Box>
  );
});

