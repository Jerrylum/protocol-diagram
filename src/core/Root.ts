import { Modals } from "./Modals";

export type RootStore = typeof rootStore;

const rootStore = {
  modals: new Modals()
} as const;

export function getRootStore(): RootStore {
  return rootStore;
}

