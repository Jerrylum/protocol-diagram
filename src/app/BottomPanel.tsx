import { Box } from "@mui/material";
import { observer } from "mobx-react-lite";
import { CommandInputField } from "./CommandInputField";

export const BottomPanel = observer(() => {
  return (
    <Box
      sx={{
        position: "fixed",
        bottom: "32px",
        maxWidth: "90vw",
        width: "600px",
        left: "50%",
        transform: "translate(-50%, 0)"
      }}>
      <CommandInputField />
    </Box>
  );
});
