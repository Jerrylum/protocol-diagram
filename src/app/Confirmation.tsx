import { Box, Button, Card, Typography } from "@mui/material";
import React from "react";
import { action, observable, reaction } from "mobx";
import { observer } from "mobx-react-lite";
import { useBetterMemo } from "../core/Hook";
import { ObserverInput } from "../component/ObserverInput";
import { Modal } from "../component/Modal";
import { getRootStore } from "../core/Root";
import { ConfirmationModalSymbol } from "../core/Confirmation";

export const ConfirmationModal = observer(() => {
  const { confirmation: cfm } = getRootStore();

  const buttons = useBetterMemo(() => observable([] as (HTMLButtonElement | null)[]), [cfm.isOpen]);
  const [renderCount, setRenderCount] = React.useState(0);

  React.useEffect(() => {
    return reaction(
      () => buttons.length,
      () => setRenderCount(renderCount => renderCount + 1)
    );
  }, [buttons]);

  React.useEffect(() => {
    (document.querySelector("*[data-testid=sentinelStart]") as HTMLElement | undefined)?.focus();
  }, [renderCount]);

  if (cfm.isOpen === false) return null;

  const focusOnButton = (offset: number) => buttons[buttons.indexOf(document.activeElement as any) + offset]?.focus();

  function onClick(idx: number) {
    if (idx < 0 || idx >= cfm.buttons.length) idx = cfm.buttons.length - 1;

    const func = cfm.buttons[idx]?.onClick;
    cfm.close();
    // ALGO: Call after closing the dialog in case the callback opens another dialog
    func?.();
  }

  function onKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (e.key === "ArrowLeft") {
      focusOnButton(-1);
    } else if (e.key === "ArrowRight") {
      focusOnButton(1);
    } else {
      const index = cfm.buttons.findIndex(btn => btn.hotkey === e.key);
      index !== -1 && onClick(index);
    }

    cfm.onKeyDown?.(e, onClick);
  }

  function onInputKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    if (e.code === "Enter" || e.code === "NumpadEnter") {
      onClick(0);
    }
  }

  return (
    <Modal symbol={ConfirmationModalSymbol} onClose={action(onClick.bind(null, -1))}>
      <Card id="confirmation-modal" className="modal-container" onKeyDown={action(onKeyDown)}>
        <Typography variant="h2" gutterBottom>
          {cfm.title}
        </Typography>
        {/* https://stackoverflow.com/questions/9769587/set-div-to-have-its-siblings-width */}
        <Box sx={{ display: "flex" }}>
          <Typography component="div" variant="body1" gutterBottom sx={{ flexGrow: "1", width: "0" }}>
            {cfm.description}
          </Typography>
        </Box>
        {cfm.input !== undefined && (
          <Box sx={{ width: "100%" }}>
            <ObserverInput
              label={cfm.inputLabel}
              getValue={() => cfm.input!}
              setValue={value => (cfm.input = value)}
              isValidIntermediate={() => true}
              isValidValue={() => true}
              onKeyDown={onInputKeyDown}
              sx={{ width: "100%" }}
            />
          </Box>
        )}
        <Box className="button-box">
          {cfm.buttons.map((btn, i) => {
            return (
              <Button
                key={i}
                disableRipple
                variant="text"
                color={btn.color ?? "inherit"}
                ref={action((element: HTMLButtonElement | null) => (buttons[i] = element))}
                onClick={action(onClick.bind(null, i))}>
                {btn.label}
                {btn.hotkey ? `(${btn.hotkey.toUpperCase()})` : ""}
              </Button>
            );
          })}
        </Box>
      </Card>
    </Modal>
  );
});
