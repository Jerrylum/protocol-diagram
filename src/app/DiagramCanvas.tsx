import { makeAutoObservable } from "mobx";
import { observer } from "mobx-react-lite";
import { Box } from "@mui/material";
import { Layer, Stage } from "react-konva";
import { Vector } from "../core/Vector";
import { getWindowSize } from "../core/Util";
import { useBetterMemo } from "../core/Hook";
import React from "react";
import Konva from "konva";

export class DiagramCanvasController {
  canvasSize: Vector = getWindowSize();

  private updateCanvasSize() {
    this.canvasSize = getWindowSize();
  }

  constructor() {
    makeAutoObservable(this);

    window.addEventListener("resize", () => this.updateCanvasSize());
  }
}

export const DiagramCanvas = observer(() => {
  const controller = useBetterMemo(() => new DiagramCanvasController(), []);
  const stageRef = React.useRef<Konva.Stage>(null);

  return (
    <Box sx={{ position: "fixed", top: 0, left: 0, bottom: 0, right: 0 }}>
      <Stage
        className="diagram-canvas"
        ref={stageRef}
        width={controller.canvasSize.x}
        height={controller.canvasSize.y}
        onContextMenu={e => e.evt.preventDefault()}>
        <Layer></Layer>
      </Stage>
    </Box>
  );
});
