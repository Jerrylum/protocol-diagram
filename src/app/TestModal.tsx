import { Card } from "@mui/material";
import { observer } from "mobx-react-lite";
import { Modal } from "../component/Modal";
import { MarkdownOverwrittenComponents } from "./MarkdownComponents";
import MarkdownMDX from "./TestMarkdown.mdx";
import { getRootStore } from "../core/Root";

export const TestModalSymbol = Symbol("TestModalSymbol");

export const TestModal = observer(() => {
  const { modals } = getRootStore();

  const onClose = () => {
    modals.close(TestModalSymbol);
  };

  return (
    <Modal symbol={TestModalSymbol} onClose={onClose}>
      <Card
        className="modal-container"
        sx={{
          padding: "16px",
          width: "768px",
          maxWidth: "80%",
          minHeight: "96px",
          outline: "none !important",
          height: "80%",
          overflowY: "auto"
        }}>
        <MarkdownMDX components={MarkdownOverwrittenComponents} />
      </Card>
    </Modal>
  );
});

