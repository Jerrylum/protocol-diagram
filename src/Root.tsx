import "./Root.scss";
import { observer } from "mobx-react-lite";
import { DiagramCanvas } from "./app/DiagramCanvas";
import { Box } from "@mui/material";
import { BottomPanel } from "./app/BottomPanel";
// import { TestModal, TestModalSymbol } from "./app/TestModal";
// import React from "react";
// import { getRootStore } from "./core/Root";

const Root = observer(() => {
  // React.useEffect(() => {
  //   getRootStore().modals.open(TestModalSymbol);
  // }, []);
  return (
    <Box id="root-container">
      <DiagramCanvas />
      <BottomPanel />
      {/* <TestModal /> */}
    </Box>
  );
});

export default Root;

