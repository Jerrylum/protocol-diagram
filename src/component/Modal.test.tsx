import { Card } from "@mui/material";
import { act, render } from "@testing-library/react";
import { TestModal, TestModalSymbol } from "../app/TestModal";
import { getRootStore } from "../core/Root";
import { Modal } from "./Modal";

test("Render TestModal", () => {
  const { modals } = getRootStore();

  const components = (
    <div id="root-container">
      <TestModal />
    </div>
  );

  const result = render(components);
  const container = document.querySelector(".modal-container");
  expect(container).not.toBeInTheDocument();

  act(() => {
    modals.close();
    modals.open(TestModalSymbol);
  });

  result.rerender(components);

  const container2 = document.querySelector(".modal-container");
  expect(container2).toBeInTheDocument();

  act(() => {
    modals.close();
  });

  result.rerender(components);

  const container3 = document.querySelector(".modal-container");
  expect(container3).not.toBeInTheDocument();
});

test("Render Modal", () => {
  const { modals } = getRootStore();

  const onOpen = jest.fn();
  const onClose = jest.fn();

  const components = (
    <div id="root-container">
      <Modal symbol={TestModalSymbol} onOpen={onOpen} onClose={onClose}>
        <Card
          className="modal-container"
          sx={{
            padding: "16px",
            width: "768px",
            maxWidth: "80%",
            minHeight: "96px",
            outline: "none !important",
            height: "80%",
            overflowY: "auto"
          }}></Card>
      </Modal>
    </div>
  );

  const result = render(components);
  expect(result.container).toMatchSnapshot();

  const container = document.querySelector(".modal-container");
  expect(container).not.toBeInTheDocument();
  expect(onOpen).not.toHaveBeenCalled();
  expect(onClose).not.toHaveBeenCalled();

  act(() => {
    modals.close();
    modals.open(TestModalSymbol);
  });

  result.rerender(components);
  expect(result.container).toMatchSnapshot();

  const container2 = document.querySelector(".modal-container");
  expect(container2).toBeInTheDocument();
  expect(onOpen).toHaveBeenCalled();
  expect(onClose).not.toHaveBeenCalled();

  act(() => {
    modals.close();
  });

  expect(onOpen).toBeCalledTimes(1);
  expect(onClose).toBeCalledTimes(1);

  act(() => {
    modals.open(TestModalSymbol);
  });

  result.rerender(components);

  expect(onOpen).toBeCalledTimes(2);
  expect(onClose).toBeCalledTimes(1);

  act(() => {
    (document.querySelector(".modal-backdrop") as HTMLElement).click();
  });

  expect(onOpen).toBeCalledTimes(2);
  expect(onClose).toBeCalledTimes(2);
});
