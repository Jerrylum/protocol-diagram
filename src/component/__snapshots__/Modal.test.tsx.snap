// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Modal 1`] = `
<div>
  <div
    id="root-container"
  />
</div>
`;

exports[`Render Modal 2`] = `
<div>
  <div
    id="root-container"
    style="overflow: hidden;"
  >
    <div
      class="MuiModal-root css-79ws1d-MuiModal-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root modal-backdrop MuiModal-backdrop modal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root modal-container css-1scpm8j-MuiPaper-root-MuiCard-root"
        tabindex="-1"
      />
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
